atomic_reactor/utils/imageutil.py

```diff
@@ -92,33 +100,76 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
         session = util.RegistrySession.create_from_config(self._conf, registry)
         return util.RegistryClient(session)
 
+    def extract_file_from_image(self, image: str, src_path: str, dst_path: str) -> None:
+        """Extract file or directory from image at src_path to dst_path
 
-# OSBS2 TBD
-
-
-def get_image_history(image):
-    # get image history with skopeo / registry api
-    return []
-
-
-def inspect_built_image():
-    # get output image final/arch specific from somewhere
-    # and call get_inspect_for_image
-    return {}
-
-
-def remove_image(image, force=False):
-    # self.tasker.remove_image(image, force=force)
-    # most likely won't be needed at all
-    return {}
-
+        :param image: str, image pullspec from which to extract
+        :param src_path: str, path inside the image that points to file or directory
+                             that will be extracted
+        :param dst_path: str, path where to export file/dir
+        """
+        cmd = ['oc', 'image', 'extract', f'{image}', '--path', f'{src_path}:{dst_path}']
 
-def tag_image(image, new_image):
-    # self.tasker.tag_image(image, new_image)
-    return True
+        try:
+            retries.run_cmd(cmd)
+        except subprocess.CalledProcessError as e:
+            logger.error("Image file extraction failed\n%s", e.output)
+            raise
 
+    def download_image_archive_tarball(self, image: str, path: str) -> None:
+        """Downloads image archive tarball to path.
 
-def get_image(image):
-    # self.tasker.get_image(image)
-    # use skopeo copy
-    return {}
+        :param image: str, image pullspec to download
+        :param path: str, path including the filename of the tarball
+        """
+        cmd = ['skopeo', 'copy', f'docker://{image}', f'docker-archive://{path}']
+        try:
+            retries.run_cmd(cmd)
+        except subprocess.CalledProcessError as e:
+            logger.error("Image archive download failed:\n%s", e.output)
+            raise
+
+    def get_uncompressed_image_layer_sizes(self, path: str) -> List[Dict[str, Union[str, int]]]:
+        """Returns data about the uncompressed image layer sizes
+
+        :param path: str path to a image archive tarball
+        :return: List[Dict[str, object]], where key is layer id and int is the size
+                                 in bytes
+        """
+        with tarfile.open(path) as tar:
+            manifest_file = tar.extractfile('manifest.json')
+            if not manifest_file:
+                raise ValueError(f'manifest.json from {path} is not a regular file')
+            manifest = json.load(manifest_file)
+            layers = manifest[0]['Layers']
```

Comment:
Should we also check that there is exactly one item in the manifest? The docker [spec](https://github.com/moby/moby/blob/master/image/spec/v1.2.md#combined-image-json--filesystem-changeset-format) says there might optionally be entries for the parent images, unfortunately it does not say if the current image will be first or last in the list

Reply:
I am not sure about this one, from quick look I can't really find anything about how to identify the item we want when there is more items. I'll try to look into it some more

Reply:
It would be also good to separate these lines into a function:

```python
        with tarfile.open(src_path) as tar:
            manifest_file = tar.extractfile('manifest.json')
            if not manifest_file:
                raise ValueError(f'manifest.json from {src_path} is not a regular file')
            manifest = json.load(manifest_file)
            layers = manifest[0]['Layers']
```

or in some other ways in order to not repeat similar logic in two methods.

Reply:
@tkdchen I am not sure separate method is necessary, especially know that one function also extracts config.json, but If you feel strongly about this I'll refactor this.

Reply:
I didn't find any additional info about having the entries for parent images, so I am not sure how to proceed. I am assuming that our images will have only one entry (I was testing with couple of them), but to avoid any surprises in the future, maybe I could document this in the code and also raise an error?

Reply:
document and raise error sounds good :+1:

Reply:
@chmeliik updated